File: rust/cloud-storage/authentication_service/src/api/webhooks/user/populate_jwt.rs
use crate::api::context::ApiContext;
use axum::{
    Json,
    extract::{self, State},
    http::StatusCode,
    response::{IntoResponse, Response},
};
use macro_middleware::auth::internal_access::ValidInternalKey;
use model::{
    authentication::webhooks::populate_jwt::{PopulateJwtWebhook, PopulateJwtWebhookResponse},
    user::UserInfoWithMacroUserId,
};

/// Populate user jwt webhook
#[tracing::instrument(skip(ctx, req, _internal_access))]
pub async fn handler(
    State(ctx): State<ApiContext>,
    _internal_access: ValidInternalKey,
    extract::Json(req): extract::Json<PopulateJwtWebhook>,
) -> Result<Response, Response> {
    let email = req.email.to_lowercase();

    tracing::trace!(email, "populating user");

    let UserInfoWithMacroUserId {
        id: user_id,
        organization_id,
        email: _,
        macro_user_id,
    } = macro_db_client::user::get::get_user_info_by_email(&ctx.db, &email)
        .await
        .map_err(|e| {
            tracing::error!(email, "failed to get user info");
            (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response()
        })?;

    Ok((
        StatusCode::OK,
        Json(PopulateJwtWebhookResponse {
            user_id,
            organization_id,
            root_macro_id: macro_user_id,
        }),
    )
        .into_response())
}

File: infra/stacks/fusionauth-instance/templates/populate_jwt.js
// Using the user and registration parameters add additional values to the jwt object.
function populate(jwt, user, _registration) {
  jwt.fusion_user_id = user.id;
  jwt.email = user.email;

  var response = fetch('{{AUTH_SERVICE_URL}}/webhooks/user/jwt', {
    method: 'POST',
    headers: {
     'x-internal-auth-key': '{{INTERNAL_SECRET}}',
     'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      email: user.email
    })
  });

    if (response.status === 200) {
    var jsonResponse = JSON.parse(response.body);
    jwt.macro_user_id = jsonResponse.user_id;

    if (jsonResponse.root_macro_id) {
      jwt.root_macro_id = jsonResponse.root_macro_id;
    }

    if(jsonResponse.organization_id){
      jwt.macro_organization_id = jsonResponse.organization_id;
    }
  } else {
    throw new Error('unable to get user info for jwt');
  }
}
